trie/witness_operators.go

```diff
@@ -0,0 +1,373 @@
+		return err
+	}
+	o.Key = keyBytesToNibbles(key)
+
+	flags := make([]byte, 1)
+	_, err = input.Read(flags)
+	if err != nil {
+		return err
+	}
+	o.HasCode = flags[0]&flagCode != 0
+	o.HasStorage = flags[0]&flagStorage != 0
+
+	if flags[0]&flagNonce != 0 {
+		var nonce uint64
+		decoder := codec.NewDecoder(input, &cbor)
+		err = decoder.Decode(&nonce)
+		if err != nil {
+			return err
+		}
+
+		o.Nonce = nonce
+	}
+
+	balance := big.NewInt(0)
+
+	if flags[0]&flagBalance != 0 {
+		var balanceBytes []byte
+		balanceBytes, err = decodeByteArray(input)
+		if err != nil {
+			return err
+		}
+
+		balance.SetBytes(balanceBytes)
+	}
+
+	o.Balance = balance
+
+	return nil
+}
+
+type OperatorCode struct {
+	Code []byte
+}
+
+func (o *OperatorCode) WriteTo(output *WitnessStatsCollector) error {
+	if err := writeOpCode(OpCode, output.WithColumn(ColumnStructure)); err != nil {
+		return err
+	}
+
+	return encodeByteArray(o.Code, output.WithColumn(ColumnCodes))
+}
+
+func (o *OperatorCode) LoadFrom(input io.Reader) error {
+	code, err := decodeByteArray(input)
+	if err != nil {
+		return err
+	}
+	o.Code = code
+	return nil
+}
+
+type OperatorBranch struct {
+	Mask uint32
+}
+
+func (o *OperatorBranch) WriteTo(output *WitnessStatsCollector) error {
+	if err := writeOpCode(OpBranch, output.WithColumn(ColumnStructure)); err != nil {
+		return err
+	}
+	encoder := codec.NewEncoder(output.WithColumn(ColumnStructure), &cbor)
+	return encoder.Encode(o.Mask)
+}
+
+func (o *OperatorBranch) LoadFrom(input io.Reader) error {
+	var mask uint32
+	decoder := codec.NewDecoder(input, &cbor)
+	if err := decoder.Decode(&mask); err != nil {
+		return err
+	}
+
+	o.Mask = mask
+
+	return nil
+}
+
+type OperatorEmptyRoot struct{}
+
+func (o *OperatorEmptyRoot) WriteTo(output *WitnessStatsCollector) error {
+	return writeOpCode(OpEmptyRoot, output.WithColumn(ColumnStructure))
+}
+
+func (o *OperatorEmptyRoot) LoadFrom(input io.Reader) error {
+	// no-op
+	return nil
+}
+
+type OperatorExtension struct {
+	Key []byte
+}
+
+func (o *OperatorExtension) WriteTo(output *WitnessStatsCollector) error {
+	if err := writeOpCode(OpExtension, output.WithColumn(ColumnStructure)); err != nil {
+		return err
+	}
+	return encodeByteArray(keyNibblesToBytes(o.Key), output.WithColumn(ColumnLeafKeys))
+}
+
+func (o *OperatorExtension) LoadFrom(input io.Reader) error {
+	key, err := decodeByteArray(input)
+	if err != nil {
+		return err
+	}
+
+	o.Key = keyBytesToNibbles(key)
+
+	return nil
+}
+
+func decodeByteArray(input io.Reader) ([]byte, error) {
+	codec := codec.NewDecoder(input, &cbor)
+
+	var buffer interface{}
+	err := codec.Decode(&buffer)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	b, ok := buffer.([]byte)
+	if !ok {
+		return []byte{}, fmt.Errorf("unexpected decode result, expected []byte, got %T", buffer)
+	}
+
+	return b, nil
+}
+
+func encodeByteArray(data []byte, output io.Writer) error {
+	encoder := codec.NewEncoder(output, &cbor)
+	return encoder.Encode(data)
```

Comment:
Perhaps use `encode.Encode(&data)` for better performance

Reply:
will it change much? the data in slices is already passed by reference, it is just the metadata that will be copied: https://play.golang.org/p/Y9kWdYdbXLx